### DISCUSSION TITLE
Directional meshes don't align with vector input

### DISCUSSION THREAD

**Discussion by JacobBumgarner** [pyvista/pyvista#1875] [Q&A] [answered]
When adding a center point and directional vector to arrows, the mesh isn't rotated appropriately. 

To demonstrate this issue, what I've done is created a new local basis based on the camera's position. The basis can seen with the red, blue, and green tubes.

When adding a cone with the starting point as the cameras position and a direction as the end point of the blue tube (camera position + first basis), the cone doesn't point in the expected direction of the blue vector. What am I doing wrong here? 

https://user-images.githubusercontent.com/70919881/143135117-5af45d42-37e3-4d52-bce5-8610669fa8db.mov

-----

**To Reproduce**

```py
import numpy as np
import pyvista as pv

def add_axes(_):
    camera = np.array([position for position in p.camera_position])
    
    A = camera[1]
    B = camera[0]
    C = camera[2]
    
    v1 = A - B
    v2 = B + C
    
    b1 = v1 / np.linalg.norm(v1)
    b2 = C
    
    v3 = np.cross(b1, b2)
    b3 = v3 / np.linalg.norm(v3)
    
    line0 = pv.Line(B, B+b1).tube(radius=0.2)
    line1 = pv.Line(B, B+b2).tube(radius=0.2)
    line2 = pv.Line(B, B+b3).tube(radius=0.2)

    p.add_mesh(line0, color='blue')
    p.add_mesh(line1, color='red')
    p.add_mesh(line2, color='green')
    
    cone = pv.Cone(center=B, direction=list(B+b1))
    p.add_mesh(cone)
    
    return

p = pv.Plotter()

arrow = pv.Arrow([1,1,1], [0,0,1])
p.add_mesh(arrow)
p.add_checkbox_button_widget(add_axes, value=False)
p.add_mesh(pv.Cube())
p.show()
```

**System Information:**

```
  Date: Tue Nov 23 16:50:51 2021 EST

                OS : Darwin
            CPU(s) : 12
           Machine : x86_64
      Architecture : 64bit
       Environment : Python
        GPU Vendor : ATI Technologies Inc.
      GPU Renderer : AMD Radeon Pro 5500M OpenGL Engine
       GPU Version : 4.1 ATI-4.6.20

  Python 3.8.8 (default, May 13 2021, 14:16:36)  [Clang 12.0.5
  (clang-1205.0.22.9)]

           pyvista : 0.32.1
               vtk : 9.1.0
             numpy : 1.20.3
           imageio : 2.11.1
           appdirs : 1.4.4
            scooby : 0.5.7
            meshio : 4.4.3
        matplotlib : 3.4.2
         pyvistaqt : 0.4.0
             PyQt5 : 5.13.2
             scipy : 1.6.3
--------------------------------------------------------------------------------
```

**Comment by banesullivan** [comment, marked as answer]
It was a bit easier for me to see the problem doing this:

<details>

```py
import numpy as np
import pyvista as pv

def add_axes(_):
    camera = np.array([position for position in p.camera_position])
    
    A = camera[1]
    B = camera[0]
    C = camera[2]
    
    v1 = A - B
    v2 = B + C
    
    b1 = v1 / np.linalg.norm(v1)
    b2 = C
    
    v3 = np.cross(b1, b2)
    b3 = v3 / np.linalg.norm(v3)
    
    line0 = pv.Line(B, B+b1).tube(radius=0.2)
    line1 = pv.Line(B, B+b2).tube(radius=0.2)
    line2 = pv.Line(B, B+b3).tube(radius=0.2)

    p.add_mesh(line0, color='blue', opacity=.75)
    p.add_mesh(line1, color='red', opacity=.75)
    p.add_mesh(line2, color='green', opacity=.75)
    
    p.add_point_labels([B, ], ['B',], point_size=10, )
    p.add_point_labels([list(B+b1), ], ['B+b1',], point_size=10)
    
    # cone = pv.Cone(center=B, direction=list(B+b1))
    # p.add_mesh(cone)
    
    return

p = pv.Plotter(notebook=0)

arrow = pv.Arrow([1,1,1], [0,0,1])
p.add_mesh(arrow)
p.add_checkbox_button_widget(add_axes, value=False)
p.add_mesh(pv.Cube())
p.show()
```

</details>

<img width="624" alt="Screen Shot 2021-11-28 at 9 59 39 AM" src="https://user-images.githubusercontent.com/22067021/143778122-046116fe-df87-4cbe-9c1e-46bc8e814491.png">


------

Your issue that you intuitively assumed that the `direction` argument of `Cone()` (and many of our other Geometric creation functions) use this direction from the given center point when that is not the case (rather no intuitively, I'll admit). This is a standard coming from VTK however. The `direction` argument is a generalized direction vector that is expected to be normalized. This means when you pass point B+b1 as the `direction`, it normalizes that vector and uses it alone for the direction of the cone not relative to the center point. To get the desired effect, you need to something like the following in your callback:

```py
    direction = np.array(np.array(B+b1)-np.array(B))/np.linalg.norm(Bb1)
    p.add_point_labels([B, ], ['B',], point_size=10, )
    p.add_point_labels([list(B+b1), ], ['B+b1',], point_size=10)
    
    # cone = pv.Cone(center=B, direction=direction)
    # This places where I like it, but maybe not where you want it
    cone = pv.Cone(center=B+b1, direction=direction)
    
    p.add_mesh(cone, opacity=0.75)
```

<img width="624" alt="Screen Shot 2021-11-28 at 10 10 35 AM" src="https://user-images.githubusercontent.com/22067021/143778494-c2231351-7509-4b7b-bc89-e1d73d526a18.png">

**Reply by JacobBumgarner** [reply]
Ohhh simple misunderstanding. Thanks for the explanation Bane, helpful as always. Cheers
